### PR TITLE
Rearchitect var args support (making it actually work) [blocks: #4329]

### PR DESCRIPTION
**Pull request by tautschnig**
Previously all the logic to access values of variable argument lists was put
into our va_arg() implementation. This required guessing what the parameters
could have been, and did not take into account the initial value passed to
va_start.

Now va_start triggers constructing an array of pointers to parameters, the first
of which is the original argument to va_start. The va_list is then just a
pointer to the first element of the array. va_arg just increments the pointer.
goto_program conversion places a side_effect_exprt where va_start was, and
goto-symex populates the array with the actual values applicable for a
particular invocation of the function.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Comment by tautschnig** [comment]
Apologies for dragging in some many reviewers, that's just because I had to add the commits from #2068 to make things also work with the approach that Visual Studio uses for var args. Just the last three commits in this PR are actually new.

**Review by smowton — APPROVED** [review]

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: ec8797e3ddd98c10e2c05886f283422cbde12633).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/102429546

**Review by peterschrammel — APPROVED** [review]

**Review by allredj — COMMENTED** [review]
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 93ba1d674b24c52b3dd0312890d74eddf0c6ff7c).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/102970072
Status will be re-evaluated on next push.
Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

  - the compatibility was already broken by an earlier merge.

**Review by allredj — COMMENTED** [review]
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 77558066454f5d78b1550318d759aee1da70430d).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/102970915
Status will be re-evaluated on next push.
Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

  - the compatibility was already broken by an earlier merge.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 00f323df1efd9f5717e210d78a8fee89bd89a4c9).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/104075038

**Review by martin-cs — APPROVED** [review]
Very little touches my areas of responsibility but those bits ae fine.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 0c5f08cd6cd09865a2e9a072c9b36ba3ff959387).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/105810019

**Review by allredj — COMMENTED** [review]
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: dfd298a1744f3891955bc217da880a63294ec7f7).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/108813088
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

File: regression/cbmc/va_list3/windows-preprocessed.desc [hunk 1]
```diff
@@ -0,0 +1,8 @@
+CORE broken-smt-backend
+windows-preprocessed.i
+--winx64
+^EXIT=0$
+^SIGNAL=0$
+^VERIFICATION SUCCESSFUL$
+--
+^warning: ignoring
```
Comment (kroening) [inline regression/cbmc/va_list3/windows-preprocessed.desc:8]: Can you add a comment, say in the .desc file, how windows-preprocessed.i was produced?

File: src/ansi-c/expr2c.cpp [hunk 2]
```diff
@@ -3625,8 +3621,8 @@ std::string expr2ct::convert_with_precedence(
       return convert_prob_uniform(src, precedence=16);
     else if(statement==ID_statement_expression)
       return convert_statement_expression(src, precedence=15);
-    else if(statement==ID_gcc_builtin_va_arg_next)
-      return convert_function(src, "gcc_builtin_va_arg_next");
+    else if(statement == ID_va_start)
+      return convert_function(src, CPROVER_PREFIX "va_start");
     else
```
Comment (kroening) [inline src/ansi-c/expr2c.cpp:3622]: Does dumping still work with that change? Meaning do we still get a C program that gcc compiles?

Comment (tautschnig) [reply src/ansi-c/expr2c.cpp:3622]: The changes (in this PR) to goto-instrument/goto_program2code.cpp take care of this.

File: src/solvers/flattening/bv_pointers.cpp [hunk 3]
```diff
@@ -254,11 +254,10 @@ bvt bv_pointerst::convert_pointer_type(const exprt &expr)
     if(op_type.id()==ID_pointer)
       return convert_bv(op);
     else if(
-      op_type.id() == ID_signedbv || op_type.id() == ID_unsignedbv ||
-      op_type.id() == ID_bool || op_type.id() == ID_c_enum ||
-      op_type.id() == ID_c_enum_tag || op_type.id() == ID_bv)
+      can_cast_type<bitvector_typet>(op_type) || op_type.id() == ID_c_enum ||
+      op_type.id() == ID_c_enum_tag)
     {
```
Comment (kroening) [inline src/solvers/flattening/bv_pointers.cpp:259]: Note that ID_bool is not a bit vector type, i.e., this case is lost.

File: src/goto-symex/symex_builtin_functions.cpp [hunk 4]
```diff
@@ -206,24 +208,44 @@ void goto_symext::symex_allocate(
...
+    binary_predicate_exprt fits_slot{
+      *parameter_size,
+      ID_le,
+      from_integer(lhs_type.get_width(), parameter_size->type())};
+
+    return if_exprt{
+      fits_slot,
+      typecast_exprt::conditional_cast(parameter_expr, void_ptr_type),
+      typecast_exprt::conditional_cast(
+        address_of_exprt{parameter_expr}, void_ptr_type)};
```
Comment (kroening) [inline src/goto-symex/symex_builtin_functions.cpp:239]: I am a bit surprised that this distinction needs to be a formula -- sure this isn't known at compile time?

File: regression/cbmc/va_list3/windows-preprocessed.desc (hunk 1 above)
Comment (tautschnig) [reply regression/cbmc/va_list3/windows-preprocessed.desc:8]: It was kind-of noted in the commit message, but I'll add a note in the .desc file as well!

File: src/goto-symex/symex_builtin_functions.cpp (hunk 4 above)
Comment (tautschnig) [reply src/goto-symex/symex_builtin_functions.cpp:239]: I believe that almost always the simplifier will be able to take care of this - but apart from that this used to construct an object that doesn't even exist at compile time, so I'm not really sure what you'd like me to do?

File: src/solvers/flattening/bv_pointers.cpp (hunk 3 above)
Comment (tautschnig) [reply src/solvers/flattening/bv_pointers.cpp:259]: Ah, thanks, fixed!

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 18cbd69c46940cf642b881e19d0cf2a98300a1ad).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/111914482

**Review by kroening — APPROVED** [review]

**Pull request merged**